### PR TITLE
feat: implement v0.51 evm module and HAS functions

### LINKED ISSUE #13181: Dagger module alteration needed for HIP 906.  Create a feature flag for enabling calls to HAS system contracts. Implement Translator and Call classes for hbarAlllowance and hbarApprove

### PR DESCRIPTION
**Pull request by lukelee-sl**
**Description**:
Third in the series of tasks for HIP-906.
- Create the v0.51 version of the evm.
- Implement `hbarAllowance` and `hbarApprove` function calls in the Hedera Account Service system contract.
- Add a feature flag for turning off calls to Hedera Account Service.

**Related issue(s)**:

Fixes #13181

**Notes for reviewer**:
xTests will be forthcoming in a future pr.
Substituting the `redirectForAccount` proxy contract byte code will also be in the next pr. 
Most of the new logic is centered around `hbarAllowanceTranslator`, `hbarAllowanceCall`, `hbarApproveTranslator` and `hbarApproveCall`

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/TestHelpers.java [hunk 1]
```diff
@@ -828,6 +829,19 @@ public static org.apache.tuweni.bytes.Bytes bytesForRedirect(final byte[] subSel
                 org.apache.tuweni.bytes.Bytes.of(subSelector));
     }
 
+    public static org.apache.tuweni.bytes.Bytes bytesForRedirectAccount(
```
Comment (david-bakin-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/TestHelpers.java:832]: Why not break with convention and add a one-line comment here and at L837 saying what these methods do? (No need IMO for full javadoc.)

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/HasSystemContractTest.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+    @AfterEach
+    void clear() {
+        frameUtils.close();
+    }
+
+    /**
+     * The unit tests for HtsSystemContract are also valid for HasSystemContract.
+     * Only add tests for unique functionality.
+     */
```
Comment (david-bakin-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/HasSystemContractTest.java:90]: Why is it that the tests over at `HtsSystemContractTest.java` are also valid for HAS?  I don't get it.

(Also, this probably shouldn't be formatted as javadoc for this method.)

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/has/HasCallAttemptTest.java [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    void setUp() {
+        callTranslators = List.of(new HbarAllowanceTranslator(), new HbarApproveTranslator());
+    }
+
+    @Test
+    void returnNullAccountIfAccountNotFound() {
+        given(nativeOperations.getAccount(numberOfLongZero(NON_SYSTEM_LONG_ZERO_ADDRESS)))
+                .willReturn(null);
+        final var input = TestHelpers.bytesForRedirectAccount(
+                HbarAllowanceTranslator.HBAR_ALLOWANCE_PROXY.selector(), NON_SYSTEM_LONG_ZERO_ADDRESS);
```
Comment (david-bakin-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/has/HasCallAttemptTest.java:67]: Say, what's the meaning of `NON_SYSTEM_LONG_ZERO_ADDRESS` which is distinct from `NON_SYSTEM_BUT_IS_LONG_ZERO_ADDRESS`?

**Review by david-bakin-sl — COMMENTED** [review]
At this point I've just checked the plumbing which looks fine.

Still to do: The 4 key classes `Hbar{Allowance|Approve}{Call|Translator}` and the tests.

File: hedera-node/hedera-app/src/xtest/java/contract/AbstractContractXTest.java [hunk 4]
```diff
@@ -122,7 +122,7 @@ public abstract class AbstractContractXTest extends AbstractXTest {
     void setUp() {
         component = DaggerContractScaffoldingComponent.factory().create(metrics, configuration(), storeMetricsService);
         callAttemptFactory = new HtsCallFactory(
-                LIVE_SYNTHETIC_IDS, addressChecks, LIVE_VERIFICATION_STRATEGIES, component.callTranslators());
+                LIVE_SYNTHETIC_IDS, addressChecks, LIVE_VERIFICATION_STRATEGIES, component.callHtsTranslators());
```
[the remark this answers is not in the document]
Comment (lukelee-sl) [reply hedera-node/hedera-app/src/xtest/java/contract/AbstractContractXTest.java:125]: yes.  next pr

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/HasSystemContractTest.java (hunk 2 above)
Comment (lukelee-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/HasSystemContractTest.java:90]: `HtsSystemContract` and `HasSystemContract` now share `AbstractNativeSystemContract` where most of the common code (which turns out to be most of the code) reside.  I didn't think replicating the unit tests from `HtsSystemContract` here would be meaningful.

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/has/HasCallAttemptTest.java (hunk 3 above)
Comment (lukelee-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/has/HasCallAttemptTest.java:67]: Just a couple of addresses.  The distinction is not really meaningful in this context.  I just needed a couple of addresses to use in the test and these were available in the helper class.

**Review by tinker-michaelj — APPROVED** [review]
LGTM, ty @lukelee-sl !

**Pull request merged**